nominal/core/_queueing.py

```diff
@@ -22,35 +35,43 @@ def put(self, item: _T_contra) -> None: ...
     def shutdown(self, immediate: bool = False) -> None: ...
 
 
-def _timed_batch(q: ReadQueue[_T], max_batch_size: int, max_batch_duration: timedelta) -> Iterable[Sequence[_T]]:
+def _timed_batch(q: ReadQueue[BatchItem], max_batch_size: int, max_batch_duration: timedelta) -> Iterable[Batch]:
     """Yield batches of items from a queue, either when the batch size is reached or the batch window expires.
 
     Will not yield empty batches.
     """
-    batch: list[_T] = []
+    batch: list[BatchItem] = []
+    oldest_timestamp: IntegralNanosecondsUTC = MAX_INT64
+    newest_timestamp: IntegralNanosecondsUTC = 0
     next_batch_time = time.time() + max_batch_duration.total_seconds()
     while True:
         now = time.time()
         try:
             item = q.get(timeout=max(0, next_batch_time - now))
             if isinstance(item, QueueShutdown):
                 if batch:
-                    yield batch
+                    yield Batch(batch, oldest_timestamp, newest_timestamp)
```

Comment:
technically this could return `Batch(batch, MAX_INT64, 0)` which would be invalid data

Reply:
actually i guess it can't since it's in `if batch`

wonder if we can make this guarantee a bit more cleanly tho..